File: packages/health_connector_hk_ios/ios/health_connector_hk_ios/Sources/health_connector_hk_ios/handlers/SystolicBloodPressureHandler.swift
import Foundation
import HealthKit

/// Handler for systolic blood pressure (instant quantity type)
///
/// **Data Type:** Systolic Blood Pressure
/// **Record Pattern:** Instant (single timestamp)
/// **Aggregation:** Average, Min, Max
/// **HealthKit Type:** HKQuantityTypeIdentifier.bloodPressureSystolic
///
/// **Note:** When writing a SystolicBloodPressureRecord on iOS, it creates
/// a standalone quantity sample, not a correlation. For complete blood
/// pressure readings (systolic + diastolic together), use BloodPressureRecord.

struct SystolicBloodPressureHandler: HealthKitQuantityHandler {
    // MARK: - HealthKitTypeHandler

    static var supportedType: HealthDataTypeDto {
        return .systolicBloodPressure
    }

    static var category: HealthKitDataCategory {
        return .quantitySample
    }

    // MARK: - HealthKitSampleHandler

    static func toDTO(_ sample: HKSample) throws -> HealthRecordDto? {
        guard let quantitySample = sample as? HKQuantitySample else {
            return nil
        }
        return quantitySample.toSystolicBloodPressureRecordDto()
    }

    static func toHealthKit(_ dto: HealthRecordDto) throws -> HKSample {
        guard let systolicDto = dto as? SystolicBloodPressureRecordDto else {
            throw HealthConnectorErrors.invalidArgument(
                message: "Expected SystolicBloodPressureRecordDto, got \(type(of: dto))"
            )
        }
        return try systolicDto.toHealthKit()
    }

    static func getSampleType() -> HKSampleType {
        return HKQuantityType.quantityType(forIdentifier: .bloodPressureSystolic)!
    }

    static func extractTimestamp(_ dto: HealthRecordDto) -> Int64 {
        guard let systolicDto = dto as? SystolicBloodPressureRecordDto else {
            return 0
        }
        return systolicDto.time
    }

    // MARK: - HealthKitQuantityHandler (Aggregation Support)

    static func supportedAggregations() -> [AggregationMetricDto] {
        return [.avg, .min, .max]
    }

    static func toStatisticsOptions(_ metric: AggregationMetricDto) -> HKStatisticsOptions {
        switch metric {
        case .avg:
            return .discreteAverage
        case .min:
            return .discreteMin
        case .max:
            return .discreteMax
        default:
            return []
        }
    }
}

File: packages/health_connector_hc_android/android/src/main/kotlin/com/phamtunglam/health_connector_hc_android/handlers/SystolicBloodPressureHandler.kt
package com.phamtunglam.health_connector_hc_android.handlers

import androidx.health.connect.client.aggregate.AggregateMetric
import androidx.health.connect.client.records.BloodPressureRecord
import androidx.health.connect.client.records.Record
import androidx.health.connect.client.units.Pressure
import com.phamtunglam.health_connector_hc_android.mappers.toDto
import com.phamtunglam.health_connector_hc_android.mappers.toHealthConnect
import com.phamtunglam.health_connector_hc_android.mappers.toSystolicDto
import com.phamtunglam.health_connector_hc_android.pigeon.AggregationMetricDto
import com.phamtunglam.health_connector_hc_android.pigeon.HealthDataTypeDto
import com.phamtunglam.health_connector_hc_android.pigeon.HealthRecordDto
import com.phamtunglam.health_connector_hc_android.pigeon.MeasurementUnitDto
import com.phamtunglam.health_connector_hc_android.pigeon.PressureDto
import com.phamtunglam.health_connector_hc_android.pigeon.PressureUnitDto
import com.phamtunglam.health_connector_hc_android.pigeon.SystolicBloodPressureRecordDto
import kotlin.reflect.KClass

/**
 * Handler for Systolic Blood Pressure health data type.
 *
 * Characteristics:
 * - Category: Instant record (single timestamp)
 * - Aggregation: Supports AVG, MIN, MAX (using BloodPressureRecord.SYSTOLIC_* metrics)
 * - Health Connect Type: BloodPressureRecord (with diastolic set to 0)
 * - Note: Writes create a BloodPressureRecord with diastolic=0
 */
object SystolicBloodPressureHandler : InstantRecordHandler, AggregationSupportingHandler {
    override val supportedType: HealthDataTypeDto = HealthDataTypeDto.SYSTOLIC_BLOOD_PRESSURE

    override fun toDto(record: Record): HealthRecordDto {
        require(record is BloodPressureRecord) {
            "Expected BloodPressureRecord, got ${record::class.simpleName}"
        }
        return record.toSystolicDto()
    }

    override fun toHealthConnect(dto: HealthRecordDto): Record {
        require(dto is SystolicBloodPressureRecordDto) {
            "Expected SystolicBloodPressureRecordDto, got ${dto::class.simpleName}"
        }
        return dto.toHealthConnect()
    }

    override fun getRecordClass(): KClass<out Record> = BloodPressureRecord::class

    override fun supportedAggregations(): List<AggregationMetricDto> {
        return listOf(
            AggregationMetricDto.AVG,
            AggregationMetricDto.MIN,
            AggregationMetricDto.MAX
        )
    }

    override fun toAggregateMetric(metric: AggregationMetricDto): AggregateMetric<*> {
        return when (metric) {
            AggregationMetricDto.AVG -> BloodPressureRecord.SYSTOLIC_AVG
            AggregationMetricDto.MIN -> BloodPressureRecord.SYSTOLIC_MIN
            AggregationMetricDto.MAX -> BloodPressureRecord.SYSTOLIC_MAX
            else -> error("Unsupported aggregation metric $metric for SystolicBloodPressure. Supported: AVG, MIN, MAX")
        }
    }

    override fun extractAggregateValue(
        aggregatedValue: Any?,
        metric: AggregationMetricDto
    ): MeasurementUnitDto {
        val pressure = aggregatedValue as? Pressure
        return pressure?.toDto() ?: PressureDto(value = 0.0, unit = PressureUnitDto.MILLIMETERS_OF_MERCURY)
    }
}
